File: certior_guard/check.py
"""Static policy analysis over the closed capability vocabulary.

Because the capability set is finite and known (:mod:`certior_guard.capabilities`)
and rules are globs over it, useful properties can be decided by exhaustive
enumeration — no solver required:

**Floor domination** (the safety invariant) — for every known capability on the
always-deny floor, and every profile in every enforcing mode (``ask``/``enforce``),
the resolved decision is ``deny``. This proves no profile or mode can open a hole
in the floor: secrets, disk wipes, remote-code-exec and exfiltration stay blocked.

**Dead rules** — a profile pattern that matches no known capability (a typo or an
aspirational rule with no emitter yet).

**Shadowed asks** — an ``ask`` pattern whose every matching capability is already
blocked, so it can never actually prompt.

This is an honest finite-model check, not a general SMT proof — but over this
vocabulary it is complete for the properties above.
"""
from __future__ import annotations

from typing import Any, Dict, List

from certior_guard.capabilities import KNOWN_CAPABILITIES, known_matching
from certior_guard.engine import resolve
from certior_guard.profiles import ALWAYS_DENY, always_denied, list_profiles

ENFORCING_MODES = ("ask", "enforce")


def check() -> Dict[str, Any]:
    """Run every check across all built-in profiles; return a structured report."""
    caps = list(KNOWN_CAPABILITIES)
    floor = [c for c in caps if always_denied(c)]

    # Floor domination: exhaustive over floor caps × profiles × enforcing modes.
    violations: List[Dict[str, str]] = []
    checks = 0
    for prof in list_profiles():
        for mode in ENFORCING_MODES:
            for cap in floor:
                checks += 1
                if resolve([cap], prof, mode)["decision"] != "deny":
                    violations.append({"profile": prof.key, "mode": mode, "capability": cap})

    profiles: List[Dict[str, Any]] = []
    for prof in list_profiles():
        dead: List[str] = []
        for pat in prof.block_patterns + prof.ask_patterns:
            if not known_matching(pat):
                dead.append(pat)
        shadowed = [pat for pat in prof.ask_patterns
                    if known_matching(pat)
                    and all(prof.blocks(c) for c in known_matching(pat))]
        profiles.append({"profile": prof.key, "dead_rules": dead, "shadowed_asks": shadowed})

    return {
        "floor_size": len(floor),
        "floor_patterns": list(ALWAYS_DENY),
        "checks": checks,
        "floor_ok": not violations,
        "violations": violations,
        "profiles": profiles,
    }

File: src/check.js
'use strict';
// Policy analysis over the closed capability vocabulary. The capability set is
// finite and rules are globs over it, so these are decided by exhaustive
// enumeration, no solver:
//   - floor domination — every always-deny capability resolves to `deny` under
//     every profile and enforcing mode (the safety invariant);
//   - dead rules — a pattern matching no known capability;
//   - shadowed asks — an `ask` pattern whose matches are all already blocked.
const { KNOWN_CAPABILITIES, knownMatching } = require('./capabilities');
const { resolve } = require('./engine');
const { ALWAYS_DENY, alwaysDenied, listProfiles } = require('./profiles');

const ENFORCING_MODES = ['ask', 'enforce'];

function check() {
  const caps = Object.keys(KNOWN_CAPABILITIES);
  const floor = caps.filter((c) => alwaysDenied(c));

  const violations = [];
  let checks = 0;
  for (const prof of listProfiles()) {
    for (const mode of ENFORCING_MODES) {
      for (const cap of floor) {
        checks += 1;
        if (resolve([cap], prof, mode).decision !== 'deny') {
          violations.push({ profile: prof.key, mode, capability: cap });
        }
      }
    }
  }

  const profiles = [];
  for (const prof of listProfiles()) {
    const dead = [];
    for (const pat of [...prof.blockPatterns, ...prof.askPatterns]) {
      if (knownMatching(pat).length === 0) dead.push(pat);
    }
    const shadowed = prof.askPatterns.filter((pat) => {
      const m = knownMatching(pat);
      return m.length > 0 && m.every((c) => prof.blocks(c));
    });
    profiles.push({ profile: prof.key, deadRules: dead, shadowedAsks: shadowed });
  }

  return {
    floorSize: floor.length,
    floorPatterns: ALWAYS_DENY.slice(),
    checks,
    floorOk: violations.length === 0,
    violations,
    profiles,
  };
}

module.exports = { check };
